### DISCUSSION TITLE
ttf2otf snippet

### DISCUSSION THREAD

**Discussion by ftCLI** [fonttools/fonttools#3143] [General]
Thanks to @justvanrossum and @anthrotype help in https://github.com/fonttools/fonttools/discussions/2836 (and thanks to @behdad for implementing qu2cu!),  I made a ttf2otf snippet.

In addition to fontTools, requires:

- [click](https://github.com/pallets/click/) to create the command line interface
- [afdko](https://github.com/adobe-type-tools/afdko) to perform optional otulines quality check with checkoutlinesufo
- [cffsubr](https://github.com/adobe-type-tools/cffsubr) to subroutinize converted fonts

**Usage**
```
Usage: ttf2otf.py [OPTIONS] INPUT_PATH

  Converts TTF fonts (or TrueType flavored woff/woff2 web fonts) to OTF fonts
  (or CFF flavored woff/woff2 web fonts).

Options:
  -t, --tolerance FLOAT RANGE   Conversion tolerance (0-3, default 1). Low
                                tolerance adds more points but keeps shapes.
                                High tolerance adds few points but may change
                                shape.  [0<=x<=3]
  --scale-upm                   Scale units-per-em of source font to 1000
                                before converting to OTF
  --keep-glyphs                 Doesn't remove NULL and CR glyphs from output
                                fonts
  --no-subr                     Do not subroutinize converted fonts with
                                cffsubr
  --check-outlines              Performs optional outline quality checks and
                                removes overlaps with checkoutlinesufo
  -out, --output-dir DIRECTORY  Specify the directory where output files are
                                to be saved. If output_dir doesn't exist, will
                                be created. If not specified, files are saved
                                to the same folder.
  --recalc-timestamp            Keep the original font 'modified' timestamp
                                (head.modified) or set it to current time. By
                                default,  original timestamp is kept.
  --no-overwrite                Overwrite existing output files or save them
                                to a new file (numbers are appended at the end
                                of file name). By default, files are
                                overwritten.
  --help                        Show this message and exit.
```

**Screenshot**
![image](https://github.com/fonttools/fonttools/assets/83063506/91818a75-7cc0-4886-8e0e-a781b6d5356e)

**Link**
https://github.com/ftCLI/TTF2OTF

**Comment by behdad** [comment]
THanks! We should probably incorporate this in fonttools.

**Reply by ftCLI** [reply]
There's still a TODO: rounding fractional coordinates (I suppose it's a pathops issue, but if a source TTF has fractional coordinates the outlines of the output font are broken).

As a temporary solution, I convert the font to a temporary OTF using T2CharstringPen, convert if back to TTF and use this last TTF as input font. But looks inelegant: https://github.com/ftCLI/ftCLI/blob/eae80490864a2e0af8f50866f4e5e94951873e9b/ftCLI/Lib/converters/ttf_to_otf.py#L65

(I removed this part of code from the snippet)

**Comment by Lorp** [comment]
What does "if a source TTF has fractional coordinates" mean?

TTF by definition cannot contain explicit fractional coordinates. However, most TTFs contain implicit fractional coordinates because of the implied on-curve point exactly half way between two consecutive off-curve points. CFF and CFF2 allow non-integer coordinates, but their representation is significantly less efficient than integers.

A 100% accurate integer qu2cu operation is possible by converting each quadratic curve into a cubic curve, _but only if the upem is doubled_. So all of the _n_ + 0.5 fractional coordinates become 2 _n_ + 1 integer coordinates. It could therefore be a useful option to double the upem in this tool.

**Reply by ftCLI** [reply]
This is what happens with a TTF of Avenir Std Book:
![image](https://github.com/fonttools/fonttools/assets/83063506/8bc2178a-6097-4350-8946-aacd296b27fc)

If the TTF is opened font in FontLab:
![image](https://github.com/fonttools/fonttools/assets/83063506/6162b456-bbe2-462d-831f-9996528e96ff)

**Reply by ftCLI** [reply]
> What does "if a source TTF has fractional coordinates" mean?
> 
> TTF by definition cannot contain explicit fractional coordinates. However, most TTFs contain implicit fractional coordinates because of the implied on-curve point exactly half way between two consecutive off-curve points. CFF and CFF2 allow non-integer coordinates, but their representation is significantly less efficient than integers.
> 
> A 100% accurate integer qu2cu operation is possible by converting each quadratic curve into a cubic curve, _but only if the upem is doubled_. So all of the _n_ + 0.5 fractional coordinates become 2 _n_ + 1 integer coordinates. It could therefore be a useful option to double the upem in this tool.

Done. I added the --scale-upm INT.

I moved the script here: https://github.com/ftCLI/TTF2OTF

**Comment by behdad** [comment]
qu2cu introduces its own fractional values...

**Comment by Lorp** [comment]
I wrote:
> A 100% accurate integer qu2cu operation is possible by converting each quadratic curve into a cubic curve, but only if the upem is doubled.

Scratch that. Coordinates of cubic control points converted from integer quadratics are fractional in general.

**Comment by ftCLI** [comment]
Workaround:

```python
if self.options.safe_mode:
    # Create a temporary OTF file with T2CharStringPen without using pathops...
    ttf2otf_converter_temp = TrueTypeToCFF(font)
    ttf2otf_converter_temp.options.charstring_source = "t2"
    ttf2otf_converter_temp.options.subroutinize = False
    ttf2otf_converter_temp.options.remove_glyphs = self.options.remove_glyphs
    temp_cff_font = ttf2otf_converter_temp.run()
    temp_cff_font.flavor = None
    temp_cff_file = makeOutputFileName(font.reader.file.name, extension=".otf", suffix="_tmp")
    temp_cff_font.save(temp_cff_file)

    # ... and convert it back to a temporary TTF file that will be used for conversion
    temp_ttf_file = makeOutputFileName(temp_cff_file, extension=".ttf")
    run_shell_command(["otf2ttf", temp_cff_file, "-e", "2.0", "-o", temp_ttf_file])
    font = TTFont(temp_ttf_file, recalcTimestamp=self.options.recalc_timestamp)
    os.remove(temp_cff_file)
    os.remove(temp_ttf_file)

def get_t2_charstrings(self) -> dict:
    """
    Get CFF charstrings using T2CharStringPen without using pathops

    :return: CFF charstrings.
    """
    charstrings = {}
    glyph_set = self.font.getGlyphSet()

    for k, v in glyph_set.items():
        # Draw the glyph with T2CharStringPen and get the charstring
        t2_pen = T2CharStringPen(v.width, glyphSet=glyph_set)
        glyph_set[k].draw(t2_pen)
        charstring = t2_pen.getCharString()
        charstrings[k] = charstring

    return charstrings
```

[ttf2otf.zip](https://github.com/fonttools/fonttools/files/11614818/ttf2otf.zip)

**Reply by anthrotype** [reply]
what's the problem with skia-pathops that you are encountering? Do you mind file an bug with a reproducer?

**Reply by ftCLI** [reply]
The problem the following: I use pathops to remove overlaps and correct contours direction. With very few fonts (the ones for which FontLab displays that warning) happensa that outlines may be broken as in the image I posted.

Where can I send you the font?

Edit. This is the portion of code where pathops fails:

```python
def get_qu2cu_charstrings(self, tolerance: float = 1.0, all_cubic: bool = True):
    charstrings = {}
    glyph_set = self.font.getGlyphSet()

    for k, v in glyph_set.items():
        # Correct contours direction and remove overlaps with pathops
        pathops_path = pathops.Path()
        pathops_pen = pathops_path.getPen(glyphSet=glyph_set)
        try:
            glyph_set[k].draw(pathops_pen)
            pathops_path.simplify()
        except TypeError:
            pass

        t2_pen = T2CharStringPen(v.width, glyphSet=glyph_set)
        qu2cu_pen = Qu2CuPen(t2_pen, max_err=tolerance, all_cubic=all_cubic, reverse_direction=False)
        pathops_path.draw(qu2cu_pen)

        charstring = t2_pen.getCharString()
        charstrings[k] = charstring

    return charstrings
```

**Reply by anthrotype** [reply]
I have now idea what Fontlab does and what they mean by that "the font contains fractional coordinates" message dialog...

We need more info to understand what's going on.
The skia-pathops (bindings for Skia PathOps C++ module) repository is at https://github.com/fonttools/skia-pathops, you can file an issue over there.

**Reply by anthrotype** [reply]
> pathops fails

how does it fail? Does it give you some error?

**Reply by ftCLI** [reply]
This:
![image](https://github.com/fonttools/fonttools/assets/83063506/e6aec549-e0e7-4368-9715-35f429903712)

Also trying to run removeOverlaps on that font fails:

```
C:\Users\Cesare\Desktop\TTF2OTF>ftcli utils ttf-remove-overlaps AvenirLTStd-Book.ttf
Traceback (most recent call last):
  File "C:\Users\Cesare\AppData\Local\Programs\Python\Python310\Scripts\ftcli-script.py", line 33, in <module>
    sys.exit(load_entry_point('font-CLI', 'console_scripts', 'ftcli')())
  File "c:\users\Cesare\onedrive - studio paghe\ftcli\ftcli\ftCLI\ftCLI.py", line 31, in main
    cli()
  File "C:\Users\Cesare\AppData\Local\Programs\Python\Python310\lib\site-packages\click\core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "C:\Users\Cesare\AppData\Local\Programs\Python\Python310\lib\site-packages\click\core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "C:\Users\Cesare\AppData\Local\Programs\Python\Python310\lib\site-packages\click\core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "C:\Users\Cesare\AppData\Local\Programs\Python\Python310\lib\site-packages\click\core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "C:\Users\Cesare\AppData\Local\Programs\Python\Python310\lib\site-packages\click\core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "C:\Users\Cesare\AppData\Local\Programs\Python\Python310\lib\site-packages\click\core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "c:\users\Cesare\onedrive - studio paghe\ftcli\ftcli\ftCLI\commands\ftcli_utils.py", line 476, in ttf_remove_overlaps
    removeOverlaps(font, removeHinting=True, ignoreErrors=ignore_errors)
  File "C:\Users\Cesare\AppData\Local\Programs\Python\Python310\lib\site-packages\fontTools\ttLib\removeOverlaps.py", line 214, in removeOverlaps
    if removeTTGlyphOverlaps(
  File "C:\Users\Cesare\AppData\Local\Programs\Python\Python310\lib\site-packages\fontTools\ttLib\removeOverlaps.py", line 141, in removeTTGlyphOverlaps
    path = skPathFromGlyph(glyphName, glyphSet)
  File "C:\Users\Cesare\AppData\Local\Programs\Python\Python310\lib\site-packages\fontTools\ttLib\removeOverlaps.py", line 34, in skPathFromGlyph
    glyphSet[glyphName].draw(pathPen)
  File "C:\Users\Cesare\AppData\Local\Programs\Python\Python310\lib\site-packages\fontTools\ttLib\ttGlyphSet.py", line 181, in draw
    glyph.draw(pen, self.glyphSet.glyfTable, offset)
  File "C:\Users\Cesare\AppData\Local\Programs\Python\Python310\lib\site-packages\fontTools\ttLib\tables\_g_l_y_f.py", line 1442, in draw
    pen.qCurveTo(*contour)
  File "src\python\pathops\_pathops.pyx", line 922, in pathops._pathops.PathPen.qCurveTo
  File "src\python\pathops\_pathops.pyx", line 929, in pathops._pathops.PathPen._qCurveToOne
TypeError: 'NoneType' object is not subscriptable
```

**Reply by anthrotype** [reply]
I believe you are stumbling on this known issue https://github.com/fonttools/skia-pathops/issues/45

**Reply by ftCLI** [reply]
Ok. Luckily the workaround works fine.

**Reply by anthrotype** [reply]
I believe https://github.com/fonttools/skia-pathops/pull/66 will fix this, please give it a go if you like

**Reply by ftCLI** [reply]
Sorry, cannot manage to install it (Windows 10 x64). I already have VisualStudio. I'll try later on a Mac.

**Reply by anthrotype** [reply]
please try upgrading to the latest skia-pathops 0.8.0 I just released https://pypi.org/project/skia-pathops/0.8.0/

**Reply by ftCLI** [reply]
Thanks! It works fine but have to dig into it. Files are now converted without problems but output files are bigger with skia-pathops 0.8.0.

TrueType files of IBM Plex Serif converted with skia-pathophs 0.8.0 installed and with skia-pathops 0.7.4 (qu2cu pen tolerance set to default value: 1 / 1000 * unitsPerEm): italic styles have the same file size (and same CRC), while uprights are about 10 kb bigger.


[skia_0_7_4.zip](https://github.com/fonttools/fonttools/files/11628022/skia_0_7_4.zip)
[skia_0_8_0.zip](https://github.com/fonttools/fonttools/files/11628024/skia_0_8_0.zip)

**Reply by anthrotype** [reply]
maybe disable subroutinization so it's easier to compare the two sets of fonts, subroutinization adds noise to the diff

**Reply by ftCLI** [reply]
Same. Italics have identical CRC also with `--no-subr` options, uprights are bigger.

[skia_0_7_4.zip](https://github.com/fonttools/fonttools/files/11628182/skia_0_7_4.zip)
[skia_0_8_0.zip](https://github.com/fonttools/fonttools/files/11628183/skia_0_8_0.zip)

**Reply by anthrotype** [reply]
strange, it appears as if Qu2Cu isn't actually being run on the outlines inside skia_0_8_0.zip.. it's as if you get one cubic per individual atomic quadratic bezier curve in the input, which is usally too many (ideally you get 1 cubic for each quadratic spline), try to tweak the qu2cu tolerance maybe.. I don't see how it might be connected to the changes I made in skia-pathops. Try to not silently pass over errors which may hide bugs

**Reply by ftCLI** [reply]
Found. If I don't catch the exception here...

```python
try:
    charstrings = self.get_qu2cu_charstrings(tolerance=self.options.tolerance, all_cubic=True)
except NotImplementedError:
    try:
        generic_info_message("All cubic set to False")
        charstrings = self.get_qu2cu_charstrings(tolerance=self.options.tolerance, all_cubic=False)
    except Exception as e:
        generic_error_message(f"An error occurred while getting qu2cu charstrings {e}")
        return
```

... and set  all_cubic to True
```python
charstrings = self.get_qu2cu_charstrings(tolerance=self.options.tolerance, all_cubic=True)
```
Raises NotImplementedError from Qu2CuPen.filterContour:
https://github.com/fonttools/fonttools/blob/02a063685f216035370bbcac74f548385233c592/Lib/fontTools/pens/qu2cuPen.py#L72

**Reply by anthrotype** [reply]
Ok, thanks. It should not be too difficult to add support for oncurve-less contours in Qu2CuPen, perhaps file an issue so it doest get lost in this long thread

**Comment by anthrotype** [comment]
I wonder, how hard would it be to make @ftCLI's ttf2otf script also support converting TrueType-flavored VF to CFF2? cc @behdad

**Reply by behdad** [reply]
Good challenge!

I think there's a CFF2 VF merger pen in cffLib somewhere we can use. I'll take a look.

**Comment by ftCLI** [comment]
I resume this topic because I think the script is ready to be added as a snippet, but I'm not opening a PR because I don't know if it's ok to use click to build the command line and cffsubr.

The workaround I found to convert oncurve-less contours to cubic is the following:

- Convert the contour to cubic with the T2CharStringPen
- Convert back the charstring back to quadratic with the Cu2QuPen
- Finally, convert the Glyph object to cubic with the Qu2CuPen

```python
    for k, v in glyph_set.items():
        width = v.width

        try:
            t2_pen = T2CharStringPen(width=width, glyphSet={k: v})
            qu2cu_pen = Qu2CuPen(t2_pen, max_err=tolerance, all_cubic=True, reverse_direction=True)
            glyph_set[k].draw(qu2cu_pen)
            qu2cu_charstrings[k] = t2_pen.getCharString()

        except NotImplementedError:
            temp_t2_pen = T2CharStringPen(width=width, glyphSet=None)
            glyph_set[k].draw(temp_t2_pen)
            t2_charstring = temp_t2_pen.getCharString()

            tt_pen = TTGlyphPen(glyphSet=None)
            cu2qu_pen = Cu2QuPen(other_pen=tt_pen, max_err=tolerance, reverse_direction=False)
            t2_charstring.draw(cu2qu_pen)
            tt_glyph = tt_pen.glyph()

            t2_pen = T2CharStringPen(width=width, glyphSet=None)
            qu2cu_pen = Qu2CuPen(t2_pen, max_err=tolerance, all_cubic=True, reverse_direction=True)
            tt_glyph.draw(pen=qu2cu_pen, glyfTable=None)
            log.info(
                f"Failed to convert glyph {k} to cubic at first attempt, but succeeded at second "
                f"attempt."
            )

        charstring = t2_pen.getCharString()
```

Here's the script: [ttf2otf.zip](https://github.com/fonttools/fonttools/files/14823864/ttf2otf.zip)

```
Usage: ttf2otf.py [OPTIONS] INPUT_PATH

  Convert TrueType flavored fonts to CFF flavored fonts.

  INPUT_PATH argument can be a file or a directory. If it is a directory, all
  the TrueType flavored fonts found in the directory will be converted.

Options:
  -r, --recursive             If INPUT_PATH is a directory, search for fonts
                              recursively in subdirectories.
  -o, --output-dir DIRECTORY  Specify a directory where the output files are
                              to be saved. If the output directory doesn't
                              exist, it will be automatically created. If not
                              specified, files will be saved to the source
                              directory.
  -w, --overwrite             Overwrite the output file if it already exists.
  --recalc-timestamp          Recalculate the font's timestamp.
  --max-err FLOAT             The maximum error allowed when converting the
                              font to TrueType.
  --new-upem INTEGER          The target UPM to scale the font to.
  --correct-contours          If the source TrueType fonts have overlapping
                              contours or incorrect directions, set this flag
                              to correct them.
  --subroutinize              Whether to subroutinize the font.
  --help                      Show this message and exit.

```
